### PR TITLE
Handle api calls with Executor

### PR DESCRIPTION
**Pull request by marfavi**

### CODE REVIEW

**Comment by marfavi** [comment]
Executor handles errors, and in cases of successful api return, transforms the response type. 

Executor is injected as an instance into repositories.

Logger is injected as an instance into Executor. All api failures are now logged properly.

Logger is no longer directly injected into any repository.

File: lib/data/repositories/v2/app_config_repository.dart [hunk 1]
```diff
@@ -1,49 +1,34 @@
...
-    final Response<AppConfig> response;
-    try {
-      response = await _api.apiV2AppconfigGet();
-    } catch (e) {
-      return Left(ClientNetworkError());
+  Environment _onSuccessfulRequest(AppConfig dto) {
+    switch (environmentTypeFromJson(dto.environmentType as String)) {
+      case EnvironmentType.production:
+        return Environment.production;
+      case EnvironmentType.localdevelopment:
```
Comment (TTA777) [inline lib/data/repositories/v2/app_config_repository.dart:20]: Please add a comment that the fall through is intended

File: lib/data/repositories/utils/executor.dart [hunk 2]
```diff
@@ -1,25 +1,30 @@
...
-    Future<Response<T>> Function() f,
-    Logger logger,
-  ) async {
-    final Response<T> response;
+  const Executor(this.logger);
+  final Logger logger;
+
+  /// Executes a network request and handles errors, including IO errors.
+  /// Returns either a [RequestError] or the transformed success type.
+  Future<Either<RequestError, Ret>> execute<Dto, Ret>(
```
Comment (TTA777) [inline lib/data/repositories/utils/executor.dart:13]: Can we call the returnType something else than Ret? Doesn't indicate it's usage very well

File: lib/data/repositories/utils/executor.dart [hunk 3]
```diff
@@ -1,25 +1,30 @@
...
+    Future<Response<Dto>> Function() request, {
+    required Ret Function(Dto dto) transformer,
```
Comment (TTA777) [inline lib/data/repositories/utils/executor.dart:15]: Why are we giving the parameters in two different ways?

**Review by TTA777 — CHANGES_REQUESTED** [review]
Minor comments, overall looks good

File: lib/data/repositories/utils/executor.dart (hunk 2 above)
Comment (marfavi) [reply lib/data/repositories/utils/executor.dart]: Called it RightType. Slightly changed doc comment.

File: lib/data/repositories/utils/executor.dart (hunk 3 above)
Comment (marfavi) [reply lib/data/repositories/utils/executor.dart]: both are now positional

**Review by jonasanker — CHANGES_REQUESTED** [review]

File: lib/data/repositories/utils/executor.dart [hunk 4]
```diff
@@ -1,25 +1,32 @@
...
+        return Right(transformer(response.body as Dto));
+      } else {
+        logger.e('API Error: (${response.statusCode}) ${response.error}');
+        return Left(RequestError.fromResponse(response));
       }
-    } catch (e) {
-      return Left(ClientNetworkError());
+    } on Exception catch (e) {
+      logger.e('HTTP Client Error: $e');
+      return Left(RequestError.clientNetworkError());
```
[the remark this answers is not in the document]
Comment (marfavi) [reply lib/data/repositories/utils/executor.dart:29]: caught specific errors related to http client

File: lib/data/repositories/utils/request_types.dart [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import 'package:chopper/chopper.dart';
+import 'package:coffeecard/base/strings.dart';
+
+/// Empty success class.
+class RequestSuccess {}
+
+class RequestError {
+  RequestError(this.message, this.code);
+  final String message;
+  final int code;
```
[the remark this answers is not in the document]
Comment (marfavi) [reply lib/data/repositories/utils/request_types.dart:10]: Renamed to RequestFailure with only one field: message. Made RequestHttpFailure that also includes status code

[the remark this answers is not in the document]
Comment (marfavi) [reply lib/data/repositories/utils/request_types.dart:10]: Now that field `code` belongs to RequestHttpFailure, it should no longer be ambigious

[the remark this answers is not in the document]
Comment (marfavi) [reply lib/data/repositories/utils/request_types.dart:13]: Reasoning behind renaming RequestError -> RequestFailure is to stick to [dart's error naming conventions](https://groups.google.com/a/dartlang.org/g/misc/c/lx9CXiV3o30/m/s5l_PwpHUGAJ).

Types named Error or Exceptions are meant to be thrown. We don't throw RequestFailure

**Pull request merged**